### ISSUE TITLE
New API using async

### ISSUE THREAD

**Issue by cberner** [cberner/fuser#164]
It would be nice to have an easier to use `Filesystem` API that also uses Rust's `async` feature

**Comment by paradigm** [comment]
[Tokio-uring 0.1.0 was just announced](https://tokio.rs/blog/2021-07-tokio-uring) and, where available, is likely to show significant performance improvements for specifically filesystem I/O that would be common with FUSE.  It may be worth consideration for this issue.  The main catch is it requires a newer Linux kernel; otherwise, a different library with a different interface (e.g. normal Tokio) is needed.

**Comment by leakingtapan** [comment]
I'm working with an API that is generated from protobuf using tonic. The generated APIs are `async` functions. I would be great if I can call into the APIs from fuser's `Filesystem` API without translating the async calls to sync calls. Is there any plan on solving this issue any time soon?

@paradigm any pointers on how to integrate with the new Tokio-uring? is that something coming out of the box that can be enable from tokio at high level with some configurations, or is that some low level API needs to be written before hand, something like read `/dev/fuse` file with io_uring?

**Comment by paradigm** [comment]
> @paradigm any pointers on how to integrate with the new Tokio-uring?

I haven't actually programmed anything in it, just read about it at a high level; I don't have enough experience with it to provide much guidance.

**Comment by messense** [comment]
There is https://github.com/3442/blown-fuse , but it's linux only.

**Comment by iwanbk** [comment]
> Check out https://github.com/Sherlock-Holo/fuse3 . Seems linux only too.

@Mivik fuse3 v0.8.1 (two days after your comment) added macos support, i haven't tried it yet tho

**Comment by cberner** [comment]
Yep. I just haven't gotten around to working on its performance. Feel free to submit PRs if you'd like to work on the async API
